Refactor changeKeys

grimxreaper opened the pull request:
In this branch we are inlining everything into the changeKeys function so that we can see all the code in one place and make high-level decisions about it that are harder to make if the code is split up into different functions found in different files.


[this description will be edited later]

grimxreaper committed:
add tests and remove comments

grimxreaper committed:
add zero in front of multiply and divide operators, pass tests

grimxreaper commented:
Would it be bad practice to type "*/" operators in the commit message instead of typing them out to save space?

grimxreaper committed:
fix double equals pressed, enable tests

grimxreaper committed:
default to correct operator, all tests pass

bryzaguy commented:
> Would it be bad practice to type "*/" operators in the commit message instead of typing them out to save space?

It would not be bad practice.

bryzaguy commented on src/App.js:
```diff
@@ -199,144 +180,98 @@ class App extends React.Component {
     tempResult = this.closeParens(tempResult) + "";
     tempResult = this.addMultiplier(tempResult) + "";
 
+    var tempResultString = tempResult;
+    const reg = /\(([0123456789/*-+.]*)\)/g;
+    var parenthesisToCalculate = tempResultString.match(reg) || [];
+    while (parenthesisToCalculate.length > 0) {
+      for (var i = 0; i < parenthesisToCalculate.length; i++) {
+        let expression = parenthesisToCalculate[i];
+        try {
+          let tempResult = round(evaluate(expression), 13);
+          tempResultString = tempResultString.replace(expression, tempResult);
+        } catch (error) {}
+      }
+      parenthesisToCalculate = tempResultString.match(reg) || [];
+    }
+    try {
+      finalResult = round(evaluate(tempResultString), 13);
+    } catch (error) {}
```
Why was this moved out of the else block? How did you gain confidence that this doesn't have any unintended consequences?

grimxreaper replied:
That's a good point, I didn't think it needed to be in an `if` and `else`, but since it may mutate code that may not show up in the tests I actually just fixed it so that it is back in an `if` and `else` block and have the tests passing. Before when I tried to preserve that structure the tests were not passing. Although to answer your question more directly, I had thought that even removing the if else wouldn't be an issue because it wouldn't hit that case here `if (matches.length > 1 && matches[0] === tempResult)` unless it was supposed to.

grimxreaper replied:
And like you said, I don't want to be changing the code's behavior in a refactor so I will keep it in this if else.

bryzaguy replied:
This is a bug fix, so far, not a refactor. You have to change behavior but it should be limited to only the code that fixes the bug.

bryzaguy reviewed (commented):
The commit diffs are looking really lean and that's awesome! Commit messages are pretty terse which is great, but some are slightly awkward when prefixing with "this commit will ...". Let me know if you don't know what I mean. Overall looks pretty nice! I'm curious if you tried testing it in your browser.

grimxreaper committed:
move run calculation code back into else

grimxreaper committed:
inline recordLastNum into onClick

grimxreaper committed:
inline setCharAt into changeKeys

grimxreaper committed:
inline addMultiplier and getLastChar into calculate

grimxreaper committed:
inline reset and backspace into onClick

grimxreaper committed:
inline calculate into onClick

grimxreaper commented:
@bryzaguy Whenever you have time, no rush:
How do I reduce cyclomatic complexity without doing function or variable extractions? I’m just a little confused right now because this is still a refactor and I know we aren’t supposed to change code, so I want to make sure that I can change code for reducing cyclomatic complexity.

bryzaguy commented:
@grimxreaper Regarding commit "combine redundant else if", how do you know it was redundant? One was just checking if the last character was a dot vs if the whole thing included one. Also, there's some new code for `result + ""`, which wasn't happening before. Why was that added?

grimxreaper commented:
> @bryzaguy  Regarding commit "combine redundant else if", how do you know it was redundant? One was just checking if the last character was a dot vs if the whole thing included one. Also, there's some new code for `result + ""`, which wasn't happening before. Why was that added?

Yes, my thought process is that if the decimal point is in the result at all (whether it is the last character of result or not) it will be covered by this 'else if' statement that I kept.

I guess I don't see the case where it would be necessary for those to be separated to specifically delineate the last character being the decimal.

The nextResult + "", was a change from before, unrelated to this commit, that I had forgotten to match to the refactoring I had done. I didn't include it in the commit message although I could have added something like "and minor change" to the end of that commit message.

bryzaguy commented:
> Yes, my thought process is that if the decimal point is in the result at all (whether it is the last character of result or not) it will be covered by this 'else if' statement that I kept.

You're right! Good job!

> The nextResult + "", was a change from before, unrelated to this commit, that I had forgotten to match to the refactoring I had done. I didn't include it in the commit message although I could have added something like "and minor change" to the end of that commit message.

I would really like to get into the habit of, once you notice you missed something, revert back and try it again. Even if it's a small change. We need to have full confidence in our refactor steps by minimizing typed code.

grimxreaper commented:
Regarding the commit "remove..." a8c in case you're wondering:

Lines 158-171 for this else if block `else if ("/*+-(".includes(result[result.length - 2]))` was not actually useful because this block was covering that condition `else if (lastCharOpenParens)`. The only time we were actually entering that block was if open parens was pressed (none of the other operators), and then we weren't entering into either of the two 'if' statements.

This
```
            if (".".includes(button)) {
              result = result + "0."; //not working
            }
```
is included in both else if statements, but the code was never actually entering this block either, in either else if statements (or in any of the else ifs, it wasn't inserting a zero before the decimal anywhere). But I am addressing this next. Wanted to make a commit to showcase this clean change first.

grimxreaper commented:
@bryzaguy is it allowed to remove code that isn't functional in a refactor? Like my 3515 commit? I also want to remove the first if block here:
```
        if ("(".includes(button)) {
          if (result.length > 1) {
            if ("0123456789".includes(result[result.length - 1])) {
              return result + button.replace("(", "*(");
            }
          }
        }
```
and all instances of 
```
          if (".".includes(button)) {
            result = "0.";
          }
```
but wanted to hear from you first and let you see what that would look like in the 3515 commit before moving forward.

grimxreaper commented on src/App.js:
```diff
@@ -139,10 +138,6 @@ class App extends React.Component {
                 result.substring(result.length - 1 + 1);
             }
           }
-        } else if ((result || "").includes(".")) {
```
This condition is already covered by the code on line 48 to 51 in the onClick function.

grimxreaper commented on src/App.js:
```diff
@@ -96,9 +95,6 @@ class App extends React.Component {
-            if ("(0123456789".includes(button)) {
-              result = result + button;
-            }
             if (".".includes(button)) {
```
Both of these conditions are already covered by the code on line 48 to 51 in the onClick function.

bryzaguy commented on src/App.js:
```diff
@@ -57,300 +59,171 @@ class App extends React.Component {
         button.key !== ")"
       ) {
+        let result = this.state.result;
+        button = button.key;
```
@grimxreaper Changing the meaning of button is not a safe edit. You need to rename the variable to a unique name inside the function before inlining it. If I'm seeing it correctly this may have introduced a breaking change where you set operate to "button.key" later. Please reset the commits and try again.

grimxreaper replied:
@bryzaguy reset all the commits?

bryzaguy replied:
@grimxreaper Reset to just before inlining changeKeys and make sure you have introduced unique variable names when inlining. `button` needs to be renamed to something unique before inlining.

bryzaguy commented on src/App.js:
```diff
@@ -57,300 +59,171 @@ class App extends React.Component {
-    this.setState({
-      result: this.state.result.slice(0, -1),
-    });
+  roundedResult = (expression) => {
```
Is this used anywhere?

grimxreaper replied:
No, but I wasn't sure if I was allowed to delete it in a refactor.

grimxreaper replied:
/safe edit

bryzaguy replied:
Should be easy to verify it is unused and safe to remove any time.

bryzaguy replied:
Looks like the right place to remove it would have been when you fixed the double equals pressed.

grimxreaper replied:
Yes I thought about removing it then, but didn't know if it was allowed, but now I know. 👍🏼

bryzaguy replied:
Well you were doing a bug fix in that case, not a refactor, so if you need to remove the usage of a method in the process it’s a good practice to find if there are other usages and remove the method if you removed the only usage. It’s also a good idea to keep your changes as small as possible, so the question is: was it important to remove the usage of this method in order to fix the bug? Look back at the “fix double equals pressed…” commit to determine.

bryzaguy reviewed (commented):
I think we may have jumped into reducing complexity too soon. Let's reset back to function inlining and try again. You're doing great!

bryzaguy commented:
> @bryzaguy is it allowed to remove code that isn't functional in a refactor? Like my 3515 commit?

It's fine to remove code if you know it's not used but not technically a refactor at that point. It's more of a safe edit and one of the first places I'd look if I noticed a change in behavior. So it's good to do it during the reduce complexity phase.

bryzaguy commented:
@grimxreaper A bug was introduced in "fix double equals pressed, enable tests", can you find it and tell me how it could have been avoided?

bryzaguy commented:
I'm unable to find the commit where you'd refactored and missed `nextResult + ""`, which commit was it?

grimxreaper commented:
> I'm unable to find the commit where you'd refactored and missed `nextResult + ""`, which commit was it?

this commit: [inline closeParens into onClick ](https://github.com/grimxreaper/js-calc/pull/19/commits/c69674b6ece5893311e654b18ba631ad1ad07681)

grimxreaper commented:
> @grimxreaper A bug was introduced in "fix double equals pressed, enable tests", can you find it and tell me how it could have been avoided?

Hmm, we added the "else" back later, went through it again and I'm not seeing it...

grimxreaper commented:
Ok I renamed the variable inside of the function changeKeys before inlining it. I renamed it to keyPressed, but since keyPressed is not recognized when inlining it, I have to set it equal to button.key inside the onClick before using it. I'm a bit confused about if this is what you meant, but I'll wait to hear from you before making any more changes so I don't risk having to reset again.

I'm sure you know this already but just for clarity,`operate` was already being set to button.key. And button.key is what was being passed in to changeKeys and called 'button'.

bryzaguy commented:
> I'll wait to hear from you before making any more changes so I don't risk having to reset again.

This works.

> I'm sure you know this already but just for clarity,`operate` was already being set to button.key. And button.key is what was being passed in to changeKeys and called 'button'.

I do know that. The line you had before of `button = button.key` changes `button` for the rest of the function so calling `setState({operate: button.key})` later will result in setting `operate` to undefined.

grimxreaper commented:
Ooh! Okay! Yikes, I see.

bryzaguy commented:
> > @grimxreaper A bug was introduced in "fix double equals pressed, enable tests", can you find it and tell me how it could have been avoided?
> 
> Hmm, we added the "else" back later, went through it again and I'm not seeing it...

Keep looking. I want this try drive home the purpose of having a predictable method for inlining and resetting commits rather than fixing in a later commit.

grimxreaper commented:
Could you let me know if the latest commit is incorrect?

bryzaguy commented:
> Could you let me know if the latest commit is incorrect?

You need to make changeKeys have a single return before inlining.

grimxreaper commented:
> > Could you let me know if the latest commit is incorrect?
> 
> You need to make changeKeys have a single return before inlining.

I got that, it's working 👍🏼

grimxreaper commented:
> > > @grimxreaper A bug was introduced in "fix double equals pressed, enable tests", can you find it and tell me how it could have been avoided?
> > 
> > 
> > Hmm, we added the "else" back later, went through it again and I'm not seeing it...
> 
> Keep looking. I want this try drive home the purpose of having a predictable method for inlining and resetting commits rather than fixing in a later commit.

So before that commit, setState was running within the else block, meaning that it was only setting done to true, and result to finalResult if it hit that `else` block. After the change, even though I added back the if and else blocks, state is being updated outside the `else` block which presents a difference, but perhaps that is why double equals wasn't working in the previous version, because it was not updating state, which is why I moved it. It's hard to check because the double equals tests weren't passing before this commit. I see why we wouldn't want to set 'done' equal to true if we are still pressing equals, but we do want the result displayed. Not sure if this is the bit that introduced a bug...but if you lmk I can keep looking.

bryzaguy commented:
> > > Could you let me know if the latest commit is incorrect?
> > 
> > 
> > You need to make changeKeys have a single return before inlining.
> 
> I got that, it's working 👍🏼

You didn’t catch all the returns and are introducing a bug. Your process needs to be air tight so you can rely on it. As in, don’t eyeball. Use the search feature to catch every return in the function. Reset the commit and try again.

bryzaguy commented:
> > > > @grimxreaper A bug was introduced in "fix double equals pressed, enable tests", can you find it and tell me how it could have been avoided?
> > > 
> > > 
> > > Hmm, we added the "else" back later, went through it again and I'm not seeing it...
> > 
> > 
> > Keep looking. I want this try drive home the purpose of having a predictable method for inlining and resetting commits rather than fixing in a later commit.
> 
> So before that commit, setState was running within the else block, meaning that it was only setting done to true, and result to finalResult if it hit that `else` block. After the change, even though I added back the if and else blocks, state is being updated outside the `else` block which presents a difference, but perhaps that is why double equals wasn't working in the previous version, because it was not updating state, which is why I moved it. It's hard to check because the double equals tests weren't passing before this commit. I see why we wouldn't want to set 'done' equal to true if we are still pressing equals, but we do want the result displayed. Not sure if this is the bit that introduced a bug...but if you lmk I can keep looking.

That’s not the bug. Your clue is that if you only moved the setState out of the else, you wouldn’t have introduced a bug.

grimxreaper commented:
> > > > > @grimxreaper A bug was introduced in "fix double equals pressed, enable tests", can you find it and tell me how it could have been avoided?
> > > > 
> > > > 
> > > > Hmm, we added the "else" back later, went through it again and I'm not seeing it...
> > > 
> > > 
> > > Keep looking. I want this try drive home the purpose of having a predictable method for inlining and resetting commits rather than fixing in a later commit.
> > 
> > 
> > So before that commit, setState was running within the else block, meaning that it was only setting done to true, and result to finalResult if it hit that `else` block. After the change, even though I added back the if and else blocks, state is being updated outside the `else` block which presents a difference, but perhaps that is why double equals wasn't working in the previous version, because it was not updating state, which is why I moved it. It's hard to check because the double equals tests weren't passing before this commit. I see why we wouldn't want to set 'done' equal to true if we are still pressing equals, but we do want the result displayed. Not sure if this is the bit that introduced a bug...but if you lmk I can keep looking.
> 
> That’s not the bug. Your clue is that if you only moved the setState out of the else, you wouldn’t have introduced a bug.

Ok, hm, what was moved out of the else was the code block that started with `var tempResultString = tempResult;`.

I checked the output for result, originalLastNum, operate, and more and don't see any difference with that code block being in or out of the `else`. There was also nothing else within that `calculate` code block that I think would be affected by that, other than the value for `finalResult`. But `finalResult` is keeping the proper value with and without the `else`.  I'll keep looking.

grimxreaper commented:
> > > > Could you let me know if the latest commit is incorrect?
> > > 
> > > 
> > > You need to make changeKeys have a single return before inlining.
> > 
> > 
> > I got that, it's working 👍🏼
> 
> You didn’t catch all the returns and are introducing a bug. Your process needs to be air tight so you can rely on it. As in, don’t eyeball. Use the search feature to catch every return in the function. Reset the commit and try again.

I'm confused, there's only one return left in the function...even with checking with the search feature.

bryzaguy commented:
> … But `finalResult` is keeping the proper value with and without the `else`…

 Are you sure about that?

bryzaguy commented on src/App.js:
```diff
@@ -60,9 +60,116 @@ class App extends React.Component {
+        if ("(".includes(keyPressed)) {
+          if (result.length > 1) {
+            if ("0123456789".includes(result[result.length - 1])) {
+              return result + keyPressed.replace("(", "*(");
```
> I'm confused, there's only one return left in the function...even with checking with the search feature.

@grimxreaper what about this?

grimxreaper replied:
I took that return out too, and it's replaced with "result =", just had to change the if's to if & else if's and all tests pass.

grimxreaper replied:
I think it will be easier to see if I make a commit showing that so it's the latest commit.

grimxreaper commented:
> > … But `finalResult` is keeping the proper value with and without the `else`…
> 
> Are you sure about that?

Yes I checked with console.logs but I will re-verify.

grimxreaper committed:
inline changeKeys in onClick

grimxreaper commented:
> > … But `finalResult` is keeping the proper value with and without the `else`…
> 
> Are you sure about that?

Ok that was incorrect.
Before making the changes in the "fix double equals pressed, enable tests" commit, `finalResult` is logging to the console as only the first result returned so for "7+7" it is returning only "14" even if equals is pressed again, it will not display anything else other than "undefined" in the console, if the console.log is placed outside the 'else' block.

So this leads me to see that `finalResult` is being updated every time during the commit "fix double equals pressed, enable tests". I'm thinking that having `finalResult` being updated in two different places is potentially introducing a bug if they aren't separated with `if` `else`.

grimxreaper commented:
<img width="466" alt="Screen Shot 2021-08-05 at 2 23 05 PM" src="https://user-images.githubusercontent.com/64915527/128401528-68a5defe-a436-4d5a-8acf-faf15fee08a8.png">
This is the output of having the if & else, and moving the setState outside of the else. And then this is the output of not having the if & else. This is what I meant by `finalResult` has the same value.

<img width="421" alt="Screen Shot 2021-08-05 at 2 25 26 PM" src="https://user-images.githubusercontent.com/64915527/128402001-c6bf02aa-b92a-47b9-8c61-40cff35f0e2e.png">

grimxreaper committed:
fix finalResult's mutation into a string back to a number

grimxreaper commented:
@bryzaguy I believe there is nothing else to fix on my end from the comments already made- except perhaps the latest commit message (wasn't sure about it), haha. Could you help me figure out next steps for the refactoring?

Maybe extracting everything between lines 63 & 174 into it's own function?

bryzaguy commented:
@grimxreaper I would like you to search for similar code, refactor it to match by introducing variables that replace the differences, and extract tiny functions that you place above your App component but below your file/module imports. Really focus on naming the functions and variables by what their goal is so you don't need to look in the function to know what it does.

grimxreaper committed:
extract ifOperatorCalcResult

grimxreaper commented:
@bryzaguy Just extracted part of the onClick, I'm guessing you won't be able to check this today so I'll keep going with more. :)
